substrate/frame/assets-freezer/src/lib.rs

```diff
@@ -0,0 +1,166 @@
+//! including its configuration trait, dispatchables, storage items, events and errors.
+//!
+//! ## Overview
+//!
+//! This pallet provides the following functionality:
+//!
+//! - Pallet hooks that implement custom logic to let `pallet-assets` know whether an balance is
+//!   frozen for an account on a given asset (see: [`pallet_assets::FrozenBalance`]).
+//! - An implementation of the fungibles [inspect][docs:inspect_freeze] and the [mutation][docs:mutate_freeze]
+//!   APIs for freezes.
+//!
+//! [docs:inspect_freeze]: `frame_support::traits::fungibles::InspectFreeze`
+//! [docs:mutate_freeze]: `frame_support::traits::fungibles::MutateFreeze`
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod impls;
+mod types;
+pub use types::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use codec::FullCodec;
+	use core::fmt::Debug;
+	use frame_support::{pallet_prelude::*, traits::VariantCount, BoundedVec};
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config + pallet_assets::Config<I> {
+		/// The overarching freeze reason.
+		type RuntimeFreezeReason: VariantCount
+			+ FullCodec
+			+ TypeInfo
+			+ PartialEq
+			+ MaxEncodedLen
+			+ Clone
+			+ Debug
+			+ 'static;
+
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The maximum number of individual freeze locks that can exist on an account at any time.
+		#[pallet::constant]
+		type MaxFreezes: Get<u32>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// Number of freezes exceed `MaxFreezes`.
+		TooManyFreezes,
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		// A reducible balance has been increased due to a freeze action.
+		AssetFrozen { who: AccountIdOf<T>, asset_id: AssetIdOf<T, I>, amount: AssetBalanceOf<T, I> },
+		// A reducible balance has been reduced due to a thaw action.
+		AssetThawed { who: AccountIdOf<T>, asset_id: AssetIdOf<T, I>, amount: AssetBalanceOf<T, I> },
+	}
+
+	/// A map that stores all the current freezes applied on an account for a given AssetId.
+	#[pallet::storage]
+	pub(super) type Freezes<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		AssetIdOf<T, I>,
+		Blake2_128Concat,
+		AccountIdOf<T>,
+		BoundedVec<IdAmount<T::RuntimeFreezeReason, AssetBalanceOf<T, I>>, T::MaxFreezes>,
+		ValueQuery,
+	>;
+
+	/// A map that stores the current reducible balance for every account on a given AssetId.
+	#[pallet::storage]
+	pub(super) type FrozenBalances<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		AssetIdOf<T, I>,
+		Blake2_128Concat,
+		AccountIdOf<T>,
+		AssetBalanceOf<T, I>,
+	>;
+}
```

Comment:
Can we have some `try-state` hooks for this storage?

Reply:
Sure! I'll start working on it right away.